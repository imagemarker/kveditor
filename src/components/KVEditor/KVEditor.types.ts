import { ReactElement } from 'react';

export interface KVEditConfigType {
  validateKey?: RegExp;
  theme?: string;
  idField?: string;
  stretchLabels?: boolean;
}
export interface KVEditorPropsType {
  items?: KVItemType[];
  rawObject?: any;
  options?: KVEditConfigType;
  onChange?: (items: any) => void;
}
export interface KVEditorType {
  (props: KVEditorPropsType): ReactElement;
}
export interface KVItemType {
  key: string;
  value: any;
  options?: KVItemOptionsType;
}
export interface KVItemOptionsType {
  immutable?: boolean; // non editable
  fixed?: boolean; // non removable
  nested?: boolean; // may include key/value list
  defaultValue?: any; // default value
}
export interface ValueEffectType {
  calculated?: any;
  validator?: any;
  parser?: any;
}
