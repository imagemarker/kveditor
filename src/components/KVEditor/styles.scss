.kv-editor {
  font-family: 'Source Sans Pro', sans-serif;
  font-weight: 400;
  margin: 3px;
  padding: 3px;
  border-radius: 3px;
  display: flex;
  flex-flow: column nowrap;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  min-width: 190px;

  &.light {
    color: #333;
    background-color: #FFF;
  }
  &.dark {
    color: #EEE;
    background-color: #444;
  }
}
