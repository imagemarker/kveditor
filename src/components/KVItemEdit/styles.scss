.kv-item-edit {
  display: flex;
  flex-flow: column nowrap;

  .row {
    display: flex;
    flex-flow: row nowrap;
    flex: 1;

    input[type=text] {
      border-radius: 3px;
      padding: 5px;
      min-width: 30px;
      width: 100%;
      font-size: 14px;
    }

    .input-key {
      flex: 0 2 auto; /* no grow, shrinks double */
      display: flex;
      margin-right: 3px;
    }

    .input-value {
      flex: 1 1 auto; /* grows, shrinks halve */
      display: flex;
      margin-right: 3px;
    }

    .input-submit {
      flex: 0 1 auto;
      cursor: pointer;
      border-radius: 3px;
      padding: 4px;
      width: 32px;
      height: 32px;
    }
  }

  .error {
    margin: 5px 0;
    text-align: left;
  }

  &.light {
    .row {
      input[type=text] {
        background-color: #FFF;
        border: solid 1px #DDD;
        color: #222;

        &::placeholder {
          color: #BBB;
        }
      }

      input:focus {
        border: solid 1px #999;
        outline: none;
      }

      .input-submit {
        background: #EEE url('icons/plus-light.svg') no-repeat center center;
        border: solid 1px #DDD;
        color: #333;

        &:hover {
          background-color: #CCC;
        }
      }
    }

    .error {
      color: #F00;
    }
  }

  &.dark {
    .row {
      input[type=text] {
        background-color: #888;
        border: solid 1px #333;
        color: #FFF;

        &::placeholder {
          color: #BBB;
        }
      }

      input:focus {
        border: solid 1px #222;
        outline: none;
      }

      .input-submit {
        background: #666 url('icons/plus-dark.svg') no-repeat center center;
        border: solid 1px #333;
        color: #CCC;

        &:hover {
          background-color: #555;
        }
      }
    }

    .error {
      color: #F66;
    }
  }
}
