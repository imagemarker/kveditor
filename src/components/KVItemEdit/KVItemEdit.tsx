import React, { useRef, useState } from 'react';
import { KVItemEditType } from '@base/types';
import './styles.scss';

const KVItemEdit: KVItemEditType = ({ dispatch, keys, editorOptions, item = { key: '', value: '' } }) => {
  const inputRef = useRef(null);
  const [kv, setKv] = useState(item);
  const [error, setError] = useState(null);

  const onKeyChange = (event: React.FormEvent<HTMLInputElement>) => {
    const key = event.currentTarget.value;
    setKv(({ value }) => ({ key, value }));
  };

  const onValueChange = (event: React.FormEvent<HTMLInputElement>) => {
    const value = event.currentTarget.value;
    setKv(({ key }) => ({ key, value }));
  };

  const setKeyError = (error: string) => {
    setError(error);
    inputRef.current.focus();
    inputRef.current.select();
  };

  const onSubmit = (event: React.FormEvent) => {
    event.preventDefault();
    const isValidKey = editorOptions.validateKey.test(kv.key);

    if (kv.key === '') {
      setKeyError('Please provide key!');
      return;
    }

    if (!isValidKey) {
      setKeyError(`Key "${kv.key}" is invalid! First char should be a letter and the rest is alphanumeric.`);
      return;
    }

    const keyExists = keys.includes(kv.key);
    if (keyExists) {
      setKeyError(`Key "${kv.key}" is already exists! Please change it.`);
      return;
    }

    dispatch({ type: 'ADD', item: kv });
    setKv({ key: '', value: '' });
    setError(null);
    inputRef.current.focus();
  };

  return (
    <form className={`kv-item-edit ${editorOptions.theme}`} onSubmit={onSubmit}>
      <div className="row">
        <div className="input-key">
          <input
            autoFocus={true}
            ref={inputRef}
            name="key"
            type="text"
            value={kv.key}
            onChange={onKeyChange}
            placeholder="key"
          />
        </div>
        <div className="input-value">
          <input name="value" type="text" value={kv.value} onChange={onValueChange} placeholder="value" />
        </div>
        <input className="input-submit" type="submit" value="Add" />
      </div>
      {error && <div className="error">{error}</div>}
    </form>
  );
};

export default KVItemEdit;
