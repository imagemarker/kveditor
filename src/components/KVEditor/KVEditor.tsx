import React, { useEffect, useReducer } from 'react';
import { KVEditorType, KVItemType } from '@base/types';
import KVReducer from '../../hooks/KVReducer';
import KVItemEdit from '../KVItemEdit/KVItemEdit';
import KVItemView from '../KVItemView/KVItemView';
import './styles.scss';

const validateKey = new RegExp(/^[a-zA-Z][a-zA-Z0-9]*$/);

const KVEditor: KVEditorType = ({
  options = { theme: 'light', validateKey, typeNotation: 'string', nested: false }
}) => {
  const [state, dispatch] = useReducer(KVReducer, { items: [], keys: [] });

  useEffect(() => {
    const items = state.items.reduce(
      (acc: Record<string, unknown>, { key, value }) => ({ ...acc, [key]: value }),
      {}
    );
    console.log(items);
  }, [state]);

  return (
    <div className={`kv-editor ${options.theme}`}>
      {state.items.map((item: KVItemType) => (
        <KVItemView
          key={item.key}
          item={item}
          dispatch={dispatch}
          editorOptions={options}
          viewOptions={null}
        />
      ))}
      <KVItemEdit keys={state.keys} dispatch={dispatch} editorOptions={options} />
    </div>
  );
};

export default KVEditor;
