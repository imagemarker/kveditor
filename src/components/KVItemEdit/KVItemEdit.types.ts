import { Dispatch, Reducer, ReducerAction } from 'react';
import { KVAction, KVState } from '@base/src/hooks/KVReducer.types';
import { KVEditConfigType, KVItemType } from '@base/src/components/KVEditor/KVEditor.types';

export interface KVItemEditPropsType {
  dispatch: Dispatch<ReducerAction<Reducer<KVState, KVAction>>>;
  keys: string[];
  editorOptions: KVEditConfigType;
  item?: KVItemType;
}
export interface KVItemEditType {
  (props: KVItemEditPropsType): JSX.Element;
}
