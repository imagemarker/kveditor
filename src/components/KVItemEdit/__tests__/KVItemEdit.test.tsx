import React from 'react';
import { render } from '@testing-library/react';
import { KVEditConfigType } from '../../KVEditor/KVEditor.types';
import KVItemEdit from '../KVItemEdit';

describe('KVItemEdit', () => {
  const validateKey = new RegExp(/^[a-zA-Z][a-zA-Z0-9]*$/);
  const dispatch = jest.fn();

  const tree = (options: KVEditConfigType = { theme: 'light', validateKey }) =>
    render(<KVItemEdit keys={['foo']} dispatch={dispatch} editorOptions={options} />);

  test('renders properly', () => {
    const { asFragment } = tree();
    expect(asFragment()).toMatchSnapshot();
  });
});
