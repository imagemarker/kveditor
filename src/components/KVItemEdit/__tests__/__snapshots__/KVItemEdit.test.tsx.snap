// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KVItemEdit renders properly 1`] = `
<DocumentFragment>
  <form
    class="kv-item-edit light"
  >
    <div
      class="row"
    >
      <div
        class="input-key"
      >
        <input
          name="key"
          placeholder="key"
          type="text"
          value=""
        />
      </div>
      <div
        class="input-value"
      >
        <input
          name="value"
          placeholder="value"
          type="text"
          value=""
        />
      </div>
      <input
        class="input-submit"
        type="submit"
        value="Add"
      />
    </div>
  </form>
</DocumentFragment>
`;
